presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrderCurveSearcher.java

```diff
@@ -114,21 +129,79 @@ public static ZOrderCurveSearcher createZOrderCurveSearcher(ZOrder zOrder, List<
         for (int index = 0; index < valueRanges.size(); index++) {
             ZValueRange range = valueRanges.get(index);
 
-            Integer minimumValue = range.getMinimumValue();
-            Integer maximumValue = range.getMaximumValue();
-
-            if (minimumValue == null) {
-                minimumValue = 0;
+            List<Integer> minimumValues = range.getMinimumValues();
+            List<Optional<Integer>> optionalMinimumValues = new ArrayList<>();
+            for (Integer minimumValue : minimumValues) {
+                if (minimumValue == null) {
+                    minimumValue = 0;
+                }
+                optionalMinimumValues.add(Optional.of(minimumValue));
             }
-            if (maximumValue == null) {
-                maximumValue = (1 << encodingBits.get(index)) - 1;
+
+            List<Integer> maximumValues = range.getMaximumValues();
+            List<Optional<Integer>> optionalMaximumValues = new ArrayList<>();
+            for (Integer maximumValue : maximumValues) {
+                if (maximumValue == null) {
+                    maximumValue = (1 << encodingBits.get(index)) - 1;
+                }
+                optionalMaximumValues.add(Optional.of(maximumValue));
             }
 
-            nonNullValues.add(new ZValueRange(Optional.of(minimumValue), Optional.of(maximumValue)));
+            nonNullValues.add(new ZValueRange(optionalMinimumValues, optionalMaximumValues));
         }
         return nonNullValues;
     }
 
+    private void permuteValueRanges(int[] counters, int[] lengths, int level, List<ZAddressRange<Long>> addressRanges, List<ZValueRange> valueRanges, RangeType rangeType)
+    {
+        if (level == counters.length) {
+            findAddressesRecursively(maxBitLength - 1, 0L, totalBitLength, addressRanges, valueRanges, rangeType, counters);
+        }
+        else {
+            for (counters[level] = 0; counters[level] < lengths[level]; counters[level]++) {
+                permuteValueRanges(counters, lengths, level + 1, addressRanges, valueRanges, rangeType);
+            }
+        }
+    }
+
+    private void combineAddressRanges(List<ZAddressRange<Long>> addressRanges)
+    {
+        if (addressRanges.isEmpty()) {
+            return;
+        }
+
+        addressRanges.sort((range1, range2) -> {
+            if (range1.equals(range2)) {
+                return 0;
+            }
+            if (range1.getMinimumAddress().equals(range2.getMinimumAddress())) {
+                return (range1.getMaximumAddress() > range2.getMaximumAddress()) ? 1 : -1;
+            }
+            return (range1.getMinimumAddress() > range2.getMinimumAddress()) ? 1 : -1;
+        });
+        List<ZAddressRange<Long>> tempRanges = addressRanges.stream().collect(ImmutableList.toImmutableList());
+
+        addressRanges.clear();
```

Comment:
`clear` might not always be the best way because the input could be immutable. How about we create a new list and return it as the combined list?

Reply:
Oh, I see! In that case, I should probably create a new list before sorting as well.